### ISSUE TITLE
False positive redundant_closure when using ref pattern in closure params

### ISSUE THREAD

**Issue by cdmistman** [rust-lang/rust-clippy#8460]
### Summary

Related to #3071 which was fixed in #7661, I found a bug when using the `ref` pattern in a closure's parameters ie `|ref foo| bar(foo)`

### Lint Name

redundant-closure

### Reproducer

I tried this code:

```rust
struct Foo;

struct Bar;

impl From<&Foo> for Bar {
    fn from(foo: &Foo) -> Self {
        Bar
    }
}

fn main() {
    // this doesn't trigger redundant-closure, as expected
    Some(Foo).map(|foo| Bar::from(&foo));
    // this triggers redundant-closure
    Some(Foo).map(|ref foo| Bar::from(foo));
    // this causes a compiler error since `From<Foo> for Bar` is not implemented
    // Some(Foo).map(Bar::from);
}
```

Note that the first statement in `main` triggers expected behavior - there is no redundant closure, and so the lint is not triggered. However, as in the issue referenced above, `redundant-closure` is triggered for the second statement, with the suggestion being the third (commented-out) statement - this is unexpected behavior, since the suggested code fails to type-check and causes a compiler error.

playground (with `allows` to isolate lint): <https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ec6a0ed2580e8fc73146abf6ada9b384>
